src/Pos/TxHistory.hs

```diff
@@ -0,0 +1,241 @@
+                                              txProcessTransaction, utxoGet, _mpLocalTxs)
+import           Pos.Types                   (Address, Block, ChainDifficulty, HeaderHash,
+                                              blockTxas, difficultyL, prevBlockL)
+import           Pos.Update                  (USHolder (..))
+import           Pos.Util                    (maybeThrow)
+
+data TxHistoryAnswer = TxHistoryAnswer
+    { taLastCachedHash :: HeaderHash
+    , taCachedNum      :: Int
+    , taCachedUtxo     :: Utxo
+    , taHistory        :: [TxHistoryEntry]
+    } deriving (Show)
+
+----------------------------------------------------------------------
+-- Deduction of history
+----------------------------------------------------------------------
+
+-- | Check if given 'Address' is one of the receivers of 'Tx'
+hasReceiver :: Tx -> Address -> Bool
+hasReceiver UnsafeTx {..} addr = any ((== addr) . txOutAddress) _txOutputs
+
+-- | Given some 'Utxo', check if given 'Address' is one of the senders of 'Tx'
+hasSender :: MonadUtxoRead m => Tx -> Address -> m Bool
+hasSender UnsafeTx {..} addr = anyM hasCorrespondingOutput $ toList _txInputs
+  where hasCorrespondingOutput txIn =
+            fmap toBool $ fmap ((== addr) . txOutAddress . toaOut) <$> utxoGet txIn
+        toBool Nothing  = False
+        toBool (Just b) = b
+
+-- | Datatype for returning info about tx history
+data TxHistoryEntry = THEntry
+    { _thTxId       :: !TxId
+    , _thTx         :: !Tx
+    , _thIsOutput   :: !Bool
+    , _thDifficulty :: !(Maybe ChainDifficulty)
+    } deriving (Show, Eq, Generic)
+
+makeLenses ''TxHistoryEntry
+
+-- | Type of monad used to deduce history
+type TxSelectorT m = UtxoStateT (MaybeT m)
+
+-- | Select transactions related to given address. `Bool` indicates
+-- whether the transaction is outgoing (i. e. is sent from given address)
+getRelatedTxs
+    :: Monad m
+    => Address
+    -> [(WithHash Tx, TxWitness, TxDistribution)]
+    -> TxSelectorT m [TxHistoryEntry]
+getRelatedTxs addr txs = fmap DL.toList $
+    lift (MaybeT $ return $ topsortTxs (view _1) txs) >>=
+    foldlM step DL.empty
+  where
+    step ls (WithHash tx txId, _wit, dist) = do
+        let isIncoming = tx `hasReceiver` addr
+        isOutgoing <- tx `hasSender` addr
+        let allToAddr = all ((== addr) . txOutAddress) $ _txOutputs tx
+            isToItself = isOutgoing && allToAddr
+        if isOutgoing || isIncoming
+            then do
+            applyTxToUtxo (WithHash tx txId) dist
+            identity %= filterUtxoByAddr addr
+
+            -- Workaround to present A to A transactions as a pair of self-canceling
+            -- transactions in history
+            let resEntry = THEntry txId tx isOutgoing Nothing
+                resList = if isToItself
+                          then DL.fromList [resEntry & thIsOutput .~ False, resEntry]
+                          else DL.singleton resEntry
```

Comment:
Maybe implement `one` for `DList` somewhere and other utilities? I see a lot of usages of `DL` specific functions.